### DISCUSSION TITLE
The replacement of `publishReplay` + `refCount` seems broken

### DISCUSSION THREAD

**Discussion by JounQin** [ReactiveX/rxjs#6438] [General]
We've written a custom operator which wraps `publishReplay` + `refCount`, and it was just working perfectly.

```ts
export const publishRef = <T>(bufferSize = 1, windowTime?: number) =>
  pipe(publishReplay<T>(bufferSize, windowTime), refCount());
```

But after v7, it is recommended to migrate to share: https://rxjs.dev/deprecations/multicasting#publishreplay

So we changed the implementation following the recommendation:

```ts
export type PublishRefConfig<T> =
  | number
  | (import('rxjs/internal/operators/share').ShareConfig<T> & {
      bufferSize?: number;
      windowTime?: number;
      timestampProvider?: TimestampProvider;
    });

export const publishRef = <T>(bufferSizeOrConfig: PublishRefConfig<T> = {}) => {
  const {
    bufferSize = 1,
    windowTime,
    timestampProvider,
    connector = () =>
      new ReplaySubject(bufferSize, windowTime, timestampProvider),
    resetOnError = false,
    resetOnComplete = false,
    resetOnRefCountZero = false,
  } = typeof bufferSizeOrConfig === 'number'
    ? ({
        bufferSize: bufferSizeOrConfig,
      } as Exclude<PublishRefConfig<T>, number>)
    : bufferSizeOrConfig;
  return share<T>({
    connector,
    resetOnError,
    resetOnComplete,
    resetOnRefCountZero,
  });
};
```

But in our Angular application, it seems working differently:

When we are consuming a watching stream only on Angular template via `async` pipe, and after navigating to other rules, the stream should be completed automatically, but the current implementation does not work as before. I don't know is our new implementation correct? Or it's just incorrect documentation.

Sorry I can't reproduce it easily because it's in our own business codes, I'll try to reproduce further, but I still want to let you know it firstly.

**Comment by voliva** [comment]
I think your example is a bit hard to follow because you have changed the signature of `publishRef`, which now can have completely different behaviours based on the parameters you pass in... why don't you try to just change the inner part of `publishRef` so that it behaves exactly as it did in v6?

Wouldn't something like this work?
```ts
export const publishRef = <T>(bufferSize = 1, windowTime?: number) =>
  share<T>({
    connector: () => new ReplaySubject(bufferSize, windowTime),
    resetOnComplete: false,
  });
```

What this does is multicast the subscription by using a replay subject, which will get reset when everyone unsubscribes, but not when the source completes. The only problem this code has, if I'm not wrong, is that `windowTime` won't work in some cases, because share will reset the subject... this could probably be solved with #6430 

Again, I think it will be easier to reason about if you keep the signature of publishRef the same as before.

**Reply by JounQin** [reply]
According to the documentation, `resetOnError: false, resetOnComplete: false, resetOnRefCountZero: false,` should be used at the same time.

![image](https://user-images.githubusercontent.com/8336744/120260596-c9cf8400-c2c8-11eb-9bdd-027942a54325.png)

Or did you mean `timestampProvider` should be removed?

The new signature is used for more flexible usage, but according to the doc it should be the same.

I'll try to keep the same signature first to see the results.

**Reply by voliva** [reply]
The documentation on the official site is a bit more clear: https://rxjs.dev/api/operators/share

In there you can see how share takes in a parameter `options: ShareConfig` which defaults to {}, and you have some examples which set just one variable.

The only thing I'm missing is what is the default of each parameter: all booleans default to true (by default, share resets on all possible scenarios), and the connector is a new plain Subject()

**Reply by voliva** [reply]
Aah but I see you are looking at a migration guide and there's exactly your pattern in there

**Reply by JounQin** [reply]
Yep, so is there anything wrong? From myself or the migration guide?

**Reply by voliva** [reply]
I think the migration guide is wrong, and I'll need more time to see for alternatives that exactly replicate the behaviour of `publishReplay + refCount`.

This sandbox here illustrates it: https://codesandbox.io/s/frosty-sun-2tv3h?file=/src/index.ts

It uses the v6 implementation + the suggested v7 migration. The test `afterUnsubscribe` shows that while the v6 implementation does unsubscribe from the source, the v7 does not. And it makes sense because `resetOnRefCountZero` is set to `false` (as per the suggested refactor).

If we change that, then this issue is solved, so I think you should give it a try as it might be enough for you:
```tsx
const publishRefRefactor = <T>(bufferSize = 1, windowTime?: number) =>
  share<T>({
    connector: () => new ReplaySubject(bufferSize, windowTime),
    resetOnComplete: true,
    resetOnError: false,
    resetOnRefCountZero: false
  });
```

But the behaviour is not exactly the same either. On the second subscription of the v6 one, you receive the latest value from the first subscription before subscribing again to the source. This doesn't happen in the v7 implementation.

I guess the problem is that `share` on v7 can't fully do that, because it just handles the reset of the subject, without caring for special cases like these. When refCount reaches 0, it just gets rid of the internal `ReplaySubject` returned by connector, disregarding any latest value (and also probably rendering the `windowTime` configuration passed to ReplaySubject ignored, useless).

**Reply by backbone87** [reply]
the recommended replacements in the migration guide and on the deprecated operator pages are indeed not equivalent replacements, but nonetheless, these are recommended because they provide the most consistent behavior.

to have a truly equivalent replacement you need to always return the same subject from the connector factory **and** use `resetOnRefCountZero: true`. I have compiled multiple ways to do this here https://codesandbox.io/s/condescending-austin-f7b8b?file=/src/index.ts

edit: I need to correct myself a little bit: the replacements i provided are indeed not 100% equivalent since they reuse the same subject across different pipelines:
```ts
const publishReplayRef6 = pipe(publishReplay(/* ... */), refCount());
const publishReplayRef7 = share({
  connector: ((subject) => () => subject)(new ReplaySubject(/* ... */),
  resetOnComplete: false,
  resetOnError: false,
});

// using the v6 operator function across multiple pipelines, has different subjects for each pipeline
const p1 = of(1).pipe(publishReplayRef6);
p1.subscribe(); // 1
p1.subscribe(); // 1, 1
const p2 = of(2).pipe(publishReplayRef6);
p2.subscribe(): // 2
p2.subscribe(): // 2, 2

// using my proposed equal v7 replacement across multiple pipelines, will use the same subject
const p1 = of(1).pipe(publishReplayRef7);
p1.subscribe(); // 1
p1.subscribe(); // 1, 1
const p2 = of(2).pipe(publishReplayRef7);
p2.subscribe(): // 1, 1, 2
p2.subscribe(): // 1, 1, 2, 2
```

to fix this you need an additional indirection step:
```ts
const publishReplayRef7Really = (source$) => {
  const subject = new ReplaySubject(/* ... */);

  return source$.pipe(
    share({ connector: () => subject, resetOnComplete: false, resetOnError: false }),
  );
}

const p1 = of(1).pipe(publishReplayRef7Really);
p1.subscribe(); // 1
p1.subscribe(); // 1, 1
const p2 = of(2).pipe(publishReplayRef7Really);
p2.subscribe(): // 2
p2.subscribe(): // 2, 2
```

**Reply by backbone87** [reply]
some additional thoughts:
you probably do not really want the old publishReplay + refCount behavior. this depends a little bit on the type of source you are dealing with (hot or cold). in the hot case, your replay would "miss out" on source emissions while no one is subscribed.

```ts
function replayHot(fn) {
  const source$ = new Subject();

  const replayed$ = source$.pipe(fn);
  let s1, s2;

  setTimeout(() => s1 = replayed$.subscribe(), 800);
  setTimeout(() => source$.next(1), 900);
  setTimeout(() => s1.unsubscribe(), 1200);

  setTimeout(() => source$.next(2), 1500);

  setTimeout(() => s2 = replayed$.subscribe(), 1800);
  setTimeout(() => source$.next(3), 2100);
  setTimeout(() => s2.unsubscribe(), 2200);
}

replayHot(pipe(publishReplay(1, 1000), refCount()));
// 1, 3

replayHot(share({
  connector: () => new ReplaySubject(1, 1000),
  resetOnComplete: false,
  resetOnError: false,
  resetOnRefCountZero: () => timer(1000),
}));
// 1, 2, 3
```

i myself do not use angular, what exactly is your use case for publishReplay + refCount? caching a value so it is immediately visible in the UI when (re)entering a view?

**Reply by JounQin** [reply]
@backbone87 

We use publishReplay + refCount for not constructing the stream twice, for `stream$ = this.api.request(...)`, if it is used twice+ in html template with `stream$ | async` or in ts, without `publishRef`, the request will be called twice+, with `publishRef`, the API request result will be cached but also it will be unsubscribed automatically when leaving the route.

**Reply by backbone87** [reply]
you could try the following, which should do what you want:
```ts
share({
  connector: () => new ReplaySubject(1),
  resetOnComplete: () => timer(cacheDuration),
  resetOnError: true, // can be left out since its the default, could also something like () => timer(cooldown)
  resetOnRefCountZero: false,
})
```

**Reply by sherlock1982** [reply]
I followed the migration guide and now ended up with a strange-working app untils I found this discussion. 
I used `publishReplay(1)+refCount` for a simple share of the last value. 
When refCount zero it should reset. 

For me the good replacement will be:

    export const publishRef = <T>() => (source: Observable<T>) => source.pipe(share({
        connector: () => new ReplaySubject(1),
        resetOnError: false,
        resetOnComplete: false,
    }));

Well resetOnError & resetOnComplete I don't think I really care (my observables theoretically doesn't error or complete) but for the sake of purity I left it.

But `resetOnRefCountZero: false` mentioned in [migration guide](https://rxjs.dev/deprecations/multicasting#publishreplay) is absolutely wrong.

**Reply by ntziolis** [reply]
Same use case as @sherlock1982, when looking at the migration guide and saw `resetOnRefCountZero: false` I got suspicious and found this thread.

The entire reason we were using `publishReplay(1)` followed by `refCount()` was to:
- ensure that the base cold observable would only get subscribed once 
- and last result being 'cached' for additional subscribers 
- All this should happen only WHILE and only while ANY subscribers are still subscribed
- Once all subscribers are unsubscribed the base observable should be unsubscribed from (route change etc.)
- Note:
  - As long as the underlying observables are things like simple web request that complete after the first value is emitted you will not run into this issue
  - However in our case the base observables are long/forever-living observables that would never complete (graphql subscriptions)

We actually migrated away from `shareReplay` for this exact reason years ago. From what I can tell, the migration guides recommendation resembles the behavior of `shareReplay` instead of `publishReplay(1)` + `refCount()`.

Since we are using this EVERYWHERE I'm keen to understand what the correct settings are for the other two parameters in order to match the previous behavior 100%:
- `resetOnError`
- `resetOnComplete`

The behavior we want to achieve is:
- After all previous subscribers are unsubscribed
- The state should be as if the observable had never been subscribed to
- The next subscriber should see the exact same behavior as the initial subscriber

Any input would be highly appreciated.

I have attached a repro that illustrate that the current recommendation does not achieve the same effect as  `publishReplay(1)` + `refCount()`:
https://stackblitz.com/edit/rxjs-1rnn3g?file=index.ts

**Reply by ntziolis** [reply]
In addition the signature of share is misleading:
- When not setting `resetOnRefCountZero`at all it behaves the same way as `resetOnRefCountZero: true`. 
- The property is optional + `boolean` and hence I would expect not specifying it to have the same effect as `false`

I would suggest to either 
- change the behavior to default to the same as when `false` is specified explicitly
- or extend the signature to call out that `true` is the default when not specified

I'm not sure if this also applies to `resetOnError` and `resetOnComplete`, but it would make sense to have a look at the implementation to ensure consistent behavior.

**Reply by NoMoreViolence** [reply]
```typescript
import { interval, Observable, ReplaySubject, share, tap } from "rxjs";

// Replacement of publishReplay(1) + refCount()
function publishRef<T>(source$: Observable<T>) {
  const replaySubject = new ReplaySubject<T>(1);
  return source$.pipe(
    share({
      connector: () => replaySubject,
      resetOnError: false,
      resetOnComplete: false,
      resetOnRefCountZero: true
    })
  );
}

const pipe$ = interval(1000).pipe(
  tap(() => {
    console.log("share pipe runed");
  }),
  publishRef
);

const { unsubscribe } = pipe$.subscribe((value) => {
  console.log(value, "1");
});

setTimeout(() => {
  unsubscribe(); 
}, 4000);
```

In that case, If the `pipe$` will not be used anymore(remove or anything), the `replaySubject` (in publishRef function) will be garbage collectible (will `complete` method be called to `replaySubject`?) ??

**Comment by mfp22** [comment]
It's been a year, but I'm finding that what I really want is not possible in a simple way and probably never has been. It's like a simple version of what TanStack Query does. 

I used to think `publishReplay(1), refCount()` was exactly what I wanted, because `shareReplay(1)` is like an immovable object once a single subscriber has existed. `publishReplay(1), refCount()`, however, will unsubscribe from the source if refCount goes to zero. So this would cancel an Angular HTTP request. This is great, and that's why I believed it was good enough. 

However, what I really want is to abstract away the fact that there are multiple subscribers, and not have multiple network requests triggered. And the behavior of a single self-completing observable is that if you subscribe again, it runs again. Obviously this is desirable, because why wouldn't you want fresh data when subscribing to an observable that everybody stopped caring about a long time ago? 

So I've been playing around with the new API, which I was reluctant to try for a long time, because it's so verbose. But it seems like I am forced to choose between 2 suboptimal behaviors: After source completes, either I have `resetOnComplete: true` where every new subscription will cause a new source subscription, and then it doesn't even update old subscribers with the results, so I get inconsistent state in my UI; OR, when all my components unsubscribe, the shared state doesn't reset even with `resetOnRefCountZero: true`, because it already completed and that forced the `refCount` to go to 0 already.

So this is my problem: When the source completes, I don't want the destination observers to complete. I want the `refCount` to stay where it is until they unsubscribe. I really do want `resetOnRefCountZero: true`, but I don't want `refCount` to go to 0 when the source completes. 

But there are no options for this in `share`. In the source code (v7+), when `source` completes, it automatically calls `dest.complete()`:

```typescript
          complete: () => {
            hasCompleted = true;
            cancelReset();
            resetConnection = handleReset(reset, resetOnComplete);
            dest.complete();
          },
```

I'm writing an article about this right now, and this isn't a very satisfying conclusion :/

I want there to be an option like `propagateCompletion = true` that I can pass `false` into to keep the observers around even if the source completes. Using that in conjunction with an observable returned from `resetOnRefCountZero` would be so nice. Almost as nice as TanStack Query.

**Reply by mfp22** [reply]
This gets me exactly what I want :)

```tsx
function shareState<T>(source$: Observable<T>) {
  const replaySubject = new ReplaySubject<T>(1);
  const sharedSource$ = merge(source$, NEVER).pipe(
    tap((result) => replaySubject.next(result)),
    share()
  );
  return using(
    () => sharedSource$.subscribe(),
    () => replaySubject
  );
}
```

I realized I had already figured these tricks out while working on StateAdapt :)

**Reply by voliva** [reply]
Yep, that's exactly it.... I was going to comment that you could just make the observable not emit a complete by concatting NEVER.

I guess this is fine, but I think that you can achieve the same just by using share with the new API instead of tap + using... wouldn't this work?:

```ts
function shareState<T>(source$: Observable<T>) {
  return merge(source$, NEVER).pipe(
    share({
      resetOnRefCountZero: true,
      connector: () => new ReplaySubject<T>(1)
    })
  );
}
```

Actually, I think your solution won't work when refCount goes to zero, because the ReplaySubject is not getting re-created, so it will keep whatever value it had on the previous call ⚠️ 

You will also need to think what happens in case of error.

**Reply by mfp22** [reply]
Yeah that's perfect. I always go for `using` when I want to do something hacky, but this is very simple and I think it's better not to replay the previous value. Thank you!
